sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java

```diff
@@ -826,25 +832,111 @@ private void channelCleanUp(AsynchronousFileChannel channel) {
+        ShareRequestConditions requestConditions = options.getRequestConditions() == null
+            ? new ShareRequestConditions() : options.getRequestConditions();
+        DownloadRetryOptions retryOptions = options.getRetryOptions() == null ? new DownloadRetryOptions()
+            : options.getRetryOptions();
+        Boolean getRangeContentMd5 = options.getRangeContentMd5();
+
+        return downloadRange(range, getRangeContentMd5, requestConditions, context)
+            .map(response -> {
+                String eTag = ModelHelper.getETag(response.getHeaders());
+                ShareFileDownloadHeaders headers = ModelHelper.transformFileDownloadHeaders(response.getHeaders());
+
+                long finalEnd;
+                if (range.getEnd() == null) {
+                    finalEnd = headers.getContentRange() == null ? headers.getContentLength()
+                        : Long.parseLong(headers.getContentRange().split("/")[1]);
+                } else {
+                    finalEnd = range.getEnd();
+                }
+
+                Flux<ByteBuffer> bufferFlux  = FluxUtil.createRetriableDownloadFlux(
+                    () -> response.getValue().timeout(TIMEOUT_VALUE),
+                    (throwable, offset) -> {
+                        if (!(throwable instanceof IOException || throwable instanceof TimeoutException)) {
+                            return Flux.error(throwable);
+                        }
+
+                        long newCount = finalEnd - (offset - range.getStart());
+
+                        /*
+                         It is possible that the network stream will throw an error after emitting all data but before
+                         completing. Issuing a retry at this stage would leave the download in a bad state with incorrect count
+                         and offset values. Because we have read the intended amount of data, we can ignore the error at the end
+                         of the stream.
+                         */
+                        if (newCount == 0) {
+                            logger.warning("Exception encountered in ReliableDownload after all data read from the network but "
+                                + "but before stream signaled completion. Returning success as all data was downloaded. "
+                                + "Exception message: " + throwable.getMessage());
+                            return Flux.empty();
+                        }
+
+                        try {
+                            return downloadRange(
+                                new ShareFileRange(offset, range.getEnd()), getRangeContentMd5,
+                                requestConditions, context).flatMapMany(r -> {
+                                    String receivedETag = ModelHelper.getETag(r.getHeaders());
+                                    if ((eTag == null && receivedETag == null)
+                                        || (eTag != null && eTag.equals(receivedETag))) {
```

Comment:
should we assume etag isn't null? What would be scenario it is null ?

Reply:
fixed